File: android/app/src/main/kotlin/ec/cacaotrace/ui/pantallas/Secado.kt
package ec.cacaotrace.ui.pantallas

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.selection.selectable
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Air
import androidx.compose.material.icons.filled.AddTask
import androidx.compose.material.icons.filled.CalendarToday
import androidx.compose.material.icons.filled.Roofing
import androidx.compose.material.icons.filled.Save
import androidx.compose.material.icons.filled.StopCircle
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material.icons.filled.WbSunny
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Card
import androidx.compose.material3.Icon
import androidx.compose.material3.ListItem
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavHostController
import ec.cacaotrace.ContenedorApp
import ec.cacaotrace.datos.bd.entidades.SecadoEntidad
import ec.cacaotrace.nucleo.modelo.MetodoSecado
import ec.cacaotrace.ui.ColoresEstado
import ec.cacaotrace.ui.comun.Aviso
import ec.cacaotrace.ui.comun.BarraSuperior
import ec.cacaotrace.ui.comun.BotonGrande
import ec.cacaotrace.ui.comun.CampoNumero
import ec.cacaotrace.ui.comun.EstadoVacio
import ec.cacaotrace.ui.comun.FilaDato
import ec.cacaotrace.ui.comun.Formato
import ec.cacaotrace.ui.comun.TarjetaSeccion
import ec.cacaotrace.ui.comun.leerNumero
import java.time.Duration
import java.time.Instant
import kotlinx.coroutines.launch

/** Guía de la prueba del puñado, para cuando no hay medidor (RF-SEC-03). */
private val pruebaPunado = listOf(
    "Suena a quebradizo" to
        "Al apretar un puñado, los granos crujen y se separan solos. Está listo.",
    "Suena sordo" to
        "Los granos se pegan entre sí y no crujen. Le falta secado.",
    "Se aplasta" to
        "El grano cede al apretarlo. Le falta bastante.",
)

/** Secado del grano (RF-SEC-01 a RF-SEC-05). */
@Composable
fun PantallaSecado(
    contenedor: ContenedorApp,
    navegacion: NavHostController,
    loteId: String,
) {
    val alcance = rememberCoroutineScope()
    var secado by remember { mutableStateOf<SecadoEntidad?>(null) }
    var recarga by remember { mutableIntStateOf(0) }
    var mostrandoJornada by remember { mutableStateOf(false) }
    var mostrandoCierre by remember { mutableStateOf(false) }

    LaunchedEffect(loteId, recarga) { secado = contenedor.lotes.secadoDe(loteId) }

    Scaffold(topBar = { BarraSuperior("Secado", navegacion) }) { relleno ->
        val s = secado
        if (s == null) {
            Column(Modifier.fillMaxSize().padding(relleno).padding(16.dp)) {
                EstadoVacio(
                    icono = Icons.Default.WbSunny,
                    titulo = "Empezar el secado",
                    explicacion = "El secado baja la humedad del grano hasta un 7 % para " +
                        "que no críe moho. Los primeros dos días son los que más agua " +
                        "sueltan: extiende en capa delgada y remueve seguido.",
                )
                MetodoSecado.entries.forEach { metodo ->
                    BotonGrande(
                        texto = metodo.etiqueta,
                        icono = when (metodo) {
                            MetodoSecado.MARQUESINA -> Icons.Default.Roofing
                            MetodoSecado.SOL -> Icons.Default.WbSunny
                            MetodoSecado.SECADOR -> Icons.Default.Air
                        },
                        modifier = Modifier.padding(vertical = 6.dp),
                    ) {
                        alcance.launch {
                            contenedor.lotes.iniciarSecado(loteId, metodo)
                            recarga++
                        }
                    }
                }
            }
            return@Scaffold
        }

        val lecturas by contenedor.lotes.observarLecturasSecado(s.id)
            .collectAsState(initial = emptyList())
        val terminado = s.fin != null
        val dias = Duration.between(s.inicio, Instant.now()).toDays() + 1

        if (mostrandoJornada) {
            DialogoJornadaSecado(
                alCerrar = { mostrandoJornada = false },
                alGuardar = { humedad, punado, espesor, remociones, moho, tempAmb, hrAmb ->
                    alcance.launch {
                        contenedor.lotes.registrarLecturaSecado(
                            secadoId = s.id,
                            loteId = loteId,
                            humedadGrano = humedad,
                            pruebaPunado = punado,
                            tempAmbiente = tempAmb,
                            hrAmbiente = hrAmb,
                            espesorCm = espesor,
                            remociones = remociones,
                            moho = moho,
                        )
                        mostrandoJornada = false
                        recarga++
                    }
                },
            )
        }

        if (mostrandoCierre) {
            DialogoCerrarSecado(
                alCerrar = { mostrandoCierre = false },
                alGuardar = { kgSeco, humedad, confirmado ->
                    alcance.launch {
                        val alertas = contenedor.lotes.cerrarSecado(
                            secadoId = s.id,
                            loteId = loteId,
                            kgSeco = kgSeco,
                            humedadFinal = humedad,
                            confirmado = confirmado,
                        )
                        // Si RN-08 bloquea, el diálogo se queda abierto para
                        // que el usuario decida a conciencia.
                        if (alertas.none { it.bloquea } || confirmado) {
                            mostrandoCierre = false
                            recarga++
                        }
                    }
                },
            )
        }

        LazyColumn(
            Modifier.fillMaxSize().padding(relleno),
            contentPadding = PaddingValues(16.dp),
        ) {
            item {
                TarjetaSeccion(
                    if (terminado) "Secado terminado" else "Día $dias de secado",
                    icono = Icons.Default.CalendarToday,
                ) {
                    Column {
                        FilaDato("Método", s.metodo.etiqueta)
                        FilaDato("Jornadas registradas", "${lecturas.size}")
                        lecturas.lastOrNull()?.humedadGrano?.let { h ->
                            FilaDato(
                                "Última humedad",
                                Formato.porcentaje(h),
                                color = if (h > 7) {
                                    ColoresEstado.atencion
                                } else {
                                    ColoresEstado.bien
                                },
                            )
                        }
                        s.kgSeco?.let { FilaDato("Grano seco", Formato.kg(it)) }
                    }
                }

                if (!terminado) {
                    BotonGrande(
                        texto = "Registrar la jornada de hoy",
                        subtitulo = "Humedad, remociones y estado del grano",
                        icono = Icons.Default.AddTask,
                    ) { mostrandoJornada = true }
                    Spacer(Modifier.height(12.dp))
                    OutlinedButton(
                        onClick = { mostrandoCierre = true },
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        Icon(Icons.Default.StopCircle, contentDescription = null)
                        Text("  Cerrar el secado")
                    }
                }
                Spacer(Modifier.height(16.dp))
            }

            items(lecturas.reversed(), key = { it.id }) { l ->
                Card(Modifier.fillMaxWidth().padding(vertical = 4.dp)) {
                    ListItem(
                        leadingContent = {
                            Icon(
                                if (l.moho) Icons.Default.Warning else Icons.Default.WbSunny,
                                contentDescription = null,
                                tint = if (l.moho) {
                                    ColoresEstado.problema
                                } else {
                                    ColoresEstado.neutro
                                },
                            )
                        },
                        headlineContent = {
                            Text(
                                l.humedadGrano?.let { "Humedad ${Formato.porcentaje(it)}" }
                                    ?: l.pruebaPunado.ifBlank { "Jornada" },
                            )
                        },
                        supportingContent = {
                            Text(
                                listOfNotNull(
                                    Formato.fechaCorta(l.fecha),
                                    "${l.remociones} remociones",
                                    l.espesorCm?.let { "capa ${Formato.numero(it)} cm" },
                                    if (l.moho) "MOHO" else null,
                                ).joinToString(" · "),
                            )
                        },
                    )
                }
            }
        }
    }
}

@Composable
private fun DialogoJornadaSecado(
    alCerrar: () -> Unit,
    alGuardar: (Double?, String, Double?, Int, Boolean, Double?, Double?) -> Unit,
) {
    var humedad by remember { mutableStateOf("") }
    var espesor by remember { mutableStateOf("") }
    var remociones by remember { mutableStateOf("3") }
    var tempAmb by remember { mutableStateOf("") }
    var hrAmb by remember { mutableStateOf("") }
    var moho by remember { mutableStateOf(false) }
    var punado by remember { mutableStateOf("") }

    AlertDialog(
        onDismissRequest = alCerrar,
        title = { Text("Jornada de secado") },
        text = {
            LazyColumn {
                item {
                    CampoNumero(
                        "Humedad del grano", humedad, { humedad = it }, unidad = "%",
                        ayuda = "Si tienes medidor. Si no, usa la prueba del puñado",
                    )
                    Text("Prueba del puñado", fontSize = 17.sp)
                    pruebaPunado.forEach { (titulo, explicacion) ->
                        Column(
                            Modifier.fillMaxWidth().selectable(
                                selected = punado == titulo,
                                onClick = { punado = titulo },
                            ).padding(vertical = 4.dp),
                        ) {
                            androidx.compose.foundation.layout.Row(
                                verticalAlignment = Alignment.CenterVertically,
                            ) {
                                RadioButton(
                                    selected = punado == titulo,
                                    onClick = { punado = titulo },
                                )
                                Text(titulo, fontSize = 16.sp)
                            }
                            Text(explicacion, fontSize = 14.sp)
                        }
                    }
                    CampoNumero(
                        "Espesor de la capa", espesor, { espesor = it }, unidad = "cm",
                        ayuda = "Los primeros días, 3-4 cm",
                    )
                    CampoNumero(
                        "Remociones del día", remociones, { remociones = it },
                        decimales = false,
                    )
                    CampoNumero("Temperatura ambiente", tempAmb, { tempAmb = it }, unidad = "°C")
                    CampoNumero("Humedad ambiente", hrAmb, { hrAmb = it }, unidad = "%")
                    androidx.compose.foundation.layout.Row(
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Switch(checked = moho, onCheckedChange = { moho = it })
                        Spacer(Modifier.height(8.dp))
                        Column {
                            Text("  Vi moho en el grano", fontSize = 16.sp)
                            Text(
                                "  Separa ahora los granos afectados: el moho da un " +
                                    "sabor que no se quita después.",
                                fontSize = 14.sp,
                            )
                        }
                    }
                }
            }
        },
        confirmButton = {
            TextButton(
                onClick = {
                    alGuardar(
                        leerNumero(humedad),
                        punado,
                        leerNumero(espesor),
                        leerNumero(remociones)?.toInt() ?: 0,
                        moho,
                        leerNumero(tempAmb),
                        leerNumero(hrAmb),
                    )
                },
            ) { Text("Guardar") }
        },
        dismissButton = { TextButton(onClick = alCerrar) { Text("Cancelar") } },
    )
}

@Composable
private fun DialogoCerrarSecado(
    alCerrar: () -> Unit,
    alGuardar: (Double, Double?, Boolean) -> Unit,
) {
    var kgSeco by remember { mutableStateOf("") }
    var humedad by remember { mutableStateOf("") }
    var confirmado by remember { mutableStateOf(false) }

    AlertDialog(
        onDismissRequest = alCerrar,
        title = { Text("Cerrar el secado") },
        text = {
            Column {
                Text(
                    "Con más del 7 % de humedad el grano cría moho en el almacén, así " +
                        "que la app te pedirá confirmación.",
                    fontSize = 15.sp,
                )
                CampoNumero("Peso del grano seco", kgSeco, { kgSeco = it }, unidad = "kg")
                CampoNumero("Humedad final", humedad, { humedad = it }, unidad = "%")
                val h = leerNumero(humedad)
                if (h != null && h > 7) {
                    Aviso(
                        titulo = "Humedad alta",
                        texto = "Por encima del 7 % el lote puede perderse en el almacén. " +
                            "Si aun así quieres almacenarlo, márcalo aquí y quedará " +
                            "registrado.",
                        color = ColoresEstado.problema,
                        accion = {
                            androidx.compose.foundation.layout.Row(
                                verticalAlignment = Alignment.CenterVertically,
                            ) {
                                Switch(
                                    checked = confirmado,
                                    onCheckedChange = { confirmado = it },
                                )
                                Text("  Almacenar de todos modos", fontSize = 15.sp)
                            }
                        },
                    )
                }
            }
        },
        confirmButton = {
            TextButton(
                onClick = {
                    val kg = leerNumero(kgSeco) ?: return@TextButton
                    alGuardar(kg, leerNumero(humedad), confirmado)
                },
            ) {
                Icon(Icons.Default.Save, contentDescription = null)
                Text(" Cerrar el secado")
            }
        },
        dismissButton = { TextButton(onClick = alCerrar) { Text("Cancelar") } },
    )
}

File: referencia/flutter-app/lib/ui/pantallas/secado.dart
/// Secado del grano (RF-SEC-01 a RF-SEC-05).
library;

import 'package:flutter/material.dart';

import '../../datos/bd/base_datos.dart';
import '../../nucleo/modelo/etapas.dart';
import '../../servicios.dart';
import '../comun/widgets.dart';
import '../tema.dart';

/// Guía de la prueba del puñado, para cuando no hay medidor (RF-SEC-03).
const List<(String, String)> pruebaPunado = [
  ('Suena a quebradizo', 'Al apretar un puñado, los granos crujen y se '
      'separan solos. Está listo.'),
  ('Suena sordo', 'Los granos se pegan entre sí y no crujen. Le falta secado.'),
  ('Se aplasta', 'El grano cede al apretarlo. Le falta bastante.'),
];

class PantallaSecado extends StatefulWidget {
  const PantallaSecado({super.key, required this.loteId});
  final String loteId;

  @override
  State<PantallaSecado> createState() => _PantallaSecadoState();
}

class _PantallaSecadoState extends State<PantallaSecado> {
  Secado? _secado;
  bool _cargando = true;

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _cargar();
  }

  Future<void> _cargar() async {
    final s = ProveedorServicios.de(context);
    final secado = await s.lotes.secadoDe(widget.loteId);
    if (mounted) {
      setState(() {
        _secado = secado;
        _cargando = false;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    final s = ProveedorServicios.de(context);
    if (_cargando) {
      return const Scaffold(body: Center(child: CircularProgressIndicator()));
    }

    if (_secado == null) {
      return Scaffold(
        appBar: AppBar(title: const Text('Secado')),
        body: ListView(
          padding: const EdgeInsets.all(16),
          children: [
            const EstadoVacio(
              icono: Icons.wb_sunny_outlined,
              titulo: 'Empezar el secado',
              explicacion:
                  'El secado baja la humedad del grano hasta un 7 % para que '
                  'no críe moho. Los primeros dos días son los que más agua '
                  'sueltan: extiende en capa delgada y remueve seguido.',
            ),
            for (final m in MetodoSecado.values)
              Padding(
                padding: const EdgeInsets.only(bottom: 12),
                child: BotonGrande(
                  texto: m.etiqueta,
                  icono: switch (m) {
                    MetodoSecado.marquesina => Icons.roofing,
                    MetodoSecado.sol => Icons.wb_sunny,
                    MetodoSecado.secador => Icons.air,
                  },
                  onPressed: () async {
                    await s.lotes
                        .iniciarSecado(loteId: widget.loteId, metodo: m);
                    _cargar();
                  },
                ),
              ),
          ],
        ),
      );
    }

    final secado = _secado!;
    final terminado = secado.fin != null;
    final dias = DateTime.now().difference(secado.inicio).inDays + 1;

    return Scaffold(
      appBar: AppBar(title: Text('Secado · ${secado.metodo.etiqueta}')),
      body: StreamBuilder<List<LecturaSecado>>(
        stream: s.lotes.observarLecturasSecado(secado.id),
        builder: (context, snapshot) {
          final lecturas = snapshot.data ?? const [];
          final ultima = lecturas.isEmpty ? null : lecturas.last;

          return ListView(
            padding: const EdgeInsets.all(16),
            children: [
              TarjetaSeccion(
                titulo: terminado ? 'Secado terminado' : 'Día $dias de secado',
                icono: Icons.calendar_today_outlined,
                hijos: [
                  FilaDato('Método', secado.metodo.etiqueta),
                  FilaDato('Jornadas registradas', '${lecturas.length}'),
                  if (ultima?.humedadGrano != null)
                    FilaDato(
                      'Última humedad',
                      '${ultima!.humedadGrano!.toStringAsFixed(1)} %',
                      color: ultima.humedadGrano! > 7
                          ? ColoresEstado.atencion
                          : ColoresEstado.bien,
                    ),
                  if (secado.kgSeco != null)
                    FilaDato('Grano seco', '${secado.kgSeco} kg'),
                ],
              ),
              if (!terminado) ...[
                BotonGrande(
                  texto: 'Registrar la jornada de hoy',
                  subtitulo: 'Humedad, remociones y estado del grano',
                  icono: Icons.add_task,
                  onPressed: () => _nuevaJornada(context, secado),
                ),
                const SizedBox(height: 12),
                OutlinedButton.icon(
                  icon: const Icon(Icons.stop_circle_outlined),
                  label: const Text('Cerrar el secado'),
                  onPressed: () => _cerrar(context, secado),
                ),
              ],
              const SizedBox(height: 16),
              if (lecturas.isNotEmpty) ...[
                Text('Jornadas', style: Theme.of(context).textTheme.titleMedium),
                for (final l in lecturas.reversed)
                  Card(
                    child: ListTile(
                      leading: Icon(
                        l.moho ? Icons.warning : Icons.wb_sunny_outlined,
                        color: l.moho ? ColoresEstado.problema : null,
                      ),
                      title: Text(l.humedadGrano == null
                          ? (l.pruebaPunado.isEmpty
                              ? 'Jornada'
                              : l.pruebaPunado)
                          : 'Humedad ${l.humedadGrano!.toStringAsFixed(1)} %'),
                      subtitle: Text([
                        _fecha(l.fecha),
                        '${l.remociones} remociones',
                        if (l.espesorCm != null) 'capa ${l.espesorCm} cm',
                        if (l.moho) 'MOHO',
                      ].join(' · ')),
                    ),
                  ),
              ],
            ],
          );
        },
      ),
    );
  }

  Future<void> _nuevaJornada(BuildContext context, Secado secado) async {
    final s = ProveedorServicios.de(context);
    final humedad = TextEditingController();
    final espesor = TextEditingController();
    final remociones = TextEditingController(text: '3');
    final tempAmb = TextEditingController();
    final hrAmb = TextEditingController();
    var moho = false;
    String? punado;

    final guardar = await showModalBottomSheet<bool>(
      context: context,
      isScrollControlled: true,
      builder: (contexto) => StatefulBuilder(
        builder: (contexto, actualizar) => Padding(
          padding: EdgeInsets.fromLTRB(
              16, 24, 16, MediaQuery.of(contexto).viewInsets.bottom + 24),
          child: SingleChildScrollView(
            child: Column(
              mainAxisSize: MainAxisSize.min,
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text('Jornada de secado',
                    style: Theme.of(contexto).textTheme.titleLarge),
                const SizedBox(height: 16),
                CampoNumero(
                  etiqueta: 'Humedad del grano',
                  controlador: humedad,
                  unidad: '%',
                  ayuda: 'Si tienes medidor. Si no, usa la prueba del puñado',
                  minimo: 0,
                  maximo: 60,
                ),
                const SizedBox(height: 8),
                const Text('Prueba del puñado',
                    style: TextStyle(
                        fontSize: 17, fontWeight: FontWeight.w600)),
                for (final p in pruebaPunado)
                  RadioListTile<String>(
                    value: p.$1,
                    // ignore: deprecated_member_use
                    groupValue: punado,
                    contentPadding: EdgeInsets.zero,
                    title: Text(p.$1),
                    subtitle: Text(p.$2, style: const TextStyle(fontSize: 14)),
                    // ignore: deprecated_member_use
                    onChanged: (v) => actualizar(() => punado = v),
                  ),
                const SizedBox(height: 8),
                CampoNumero(
                  etiqueta: 'Espesor de la capa',
                  controlador: espesor,
                  unidad: 'cm',
                  ayuda: 'Los primeros días, 3-4 cm',
                  minimo: 0,
                ),
                CampoNumero(
                  etiqueta: 'Remociones del día',
                  controlador: remociones,
                  decimales: false,
                ),
                CampoNumero(
                  etiqueta: 'Temperatura ambiente',
                  controlador: tempAmb,
                  unidad: '°C',
                ),
                CampoNumero(
                  etiqueta: 'Humedad ambiente',
                  controlador: hrAmb,
                  unidad: '%',
                  maximo: 100,
                ),
                SwitchListTile(
                  value: moho,
                  contentPadding: EdgeInsets.zero,
                  title: const Text('Vi moho en el grano'),
                  subtitle: const Text(
                    'Separa ahora los granos afectados: el moho da un sabor '
                    'que no se quita después.',
                    style: TextStyle(fontSize: 14),
                  ),
                  onChanged: (v) => actualizar(() => moho = v),
                ),
                const SizedBox(height: 16),
                BotonGrande(
                  texto: 'Guardar la jornada',
                  icono: Icons.save_outlined,
                  onPressed: () => Navigator.pop(contexto, true),
                ),
              ],
            ),
          ),
        ),
      ),
    );

    if (guardar != true || !context.mounted) return;
    final alertas = await s.lotes.registrarLecturaSecado(
      secadoId: secado.id,
      loteId: widget.loteId,
      humedadGrano: leerNumero(humedad),
      pruebaPunado: punado ?? '',
      tempAmbiente: leerNumero(tempAmb),
      hrAmbiente: leerNumero(hrAmb),
      espesorCm: leerNumero(espesor),
      remociones: leerNumero(remociones)?.round() ?? 0,
      moho: moho,
    );
    if (!context.mounted) return;
    avisar(
      context,
      alertas.isEmpty
          ? 'Jornada guardada'
          : '${alertas.first.quePaso}. ${alertas.first.queHacer}',
      error: alertas.isNotEmpty,
    );
  }

  Future<void> _cerrar(BuildContext context, Secado secado) async {
    final s = ProveedorServicios.de(context);
    final kgSeco = TextEditingController();
    final humedad = TextEditingController();

    final guardar = await showModalBottomSheet<bool>(
      context: context,
      isScrollControlled: true,
      builder: (contexto) => Padding(
        padding: EdgeInsets.fromLTRB(
            16, 24, 16, MediaQuery.of(contexto).viewInsets.bottom + 24),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text('Cerrar el secado',
                style: Theme.of(contexto).textTheme.titleLarge),
            const SizedBox(height: 8),
            const Text(
              'Con más del 7 % de humedad el grano cría moho en el almacén, '
              'así que la app te pedirá confirmación.',
              style: TextStyle(fontSize: 15),
            ),
            const SizedBox(height: 16),
            CampoNumero(
              etiqueta: 'Peso del grano seco',
              controlador: kgSeco,
              unidad: 'kg',
              obligatorio: true,
              minimo: 0,
            ),
            CampoNumero(
              etiqueta: 'Humedad final',
              controlador: humedad,
              unidad: '%',
              minimo: 0,
              maximo: 60,
            ),
            const SizedBox(height: 16),
            BotonGrande(
              texto: 'Cerrar el secado',
              icono: Icons.check,
              onPressed: () => Navigator.pop(contexto, true),
            ),
          ],
        ),
      ),
    );

    if (guardar != true || !context.mounted) return;
    final kg = leerNumero(kgSeco);
    if (kg == null) {
      avisar(context, 'Escribe el peso del grano seco', error: true);
      return;
    }

    var alertas = await s.lotes.cerrarSecado(
      secadoId: secado.id,
      loteId: widget.loteId,
      kgSeco: kg,
      humedadFinal: leerNumero(humedad),
    );

    final bloqueo = alertas.where((a) => a.bloquea).firstOrNull;
    if (bloqueo != null && context.mounted) {
      final seguir = await confirmar(
        context,
        titulo: bloqueo.quePaso,
        mensaje: '${bloqueo.porQueImporta}\n\n'
            '¿Quieres almacenarlo de todos modos? Quedará registrado.',
        si: 'Almacenar igual',
        peligroso: true,
      );
      if (!seguir) return;
      alertas = await s.lotes.cerrarSecado(
        secadoId: secado.id,
        loteId: widget.loteId,
        kgSeco: kg,
        humedadFinal: leerNumero(humedad),
        confirmado: true,
      );
    }

    if (!context.mounted) return;
    avisar(context, 'Secado cerrado con $kg kg');
    _cargar();
  }
}

String _fecha(DateTime f) =>
    '${f.day.toString().padLeft(2, '0')}/'
    '${f.month.toString().padLeft(2, '0')}';
